build/Vessel.js

```diff
@@ -754,35 +754,39 @@ Object.assign(Ship.prototype, {
 		let GMl = KB + BMl - KG;
 		return {GMt, GMl, KB, BMt, BMl, KG};
 	},
-	getFuelMass() {
+	getFuelMass: function(shipState) {
+		shipState = shipState || this.designState;
+
 		var fuelMass = {};
 		fuelMass.totalMass = 0;
 		fuelMass.tankMass = {};
 		fuelMass.tankStates = {};
 		for (let o of Object.values(this.derivedObjects)) {
-			if (o.baseObject.capabilities.fuelTank) {
-				fuelMass.tankStates[o.id] = this.designState.getObjectState(o);
+			if (o.group === "fuel tanks") {
+				fuelMass.tankStates[o.id] = shipState.getObjectState(o);
 				fuelMass.tankMass[o.id] = o.baseObject.weightInformation.contentDensity * o.baseObject.weightInformation.volumeCapacity * fuelMass.tankStates[o.id].fullness;
 				fuelMass.totalMass += fuelMass.tankMass[o.id];
 			}
 		}
 		return fuelMass;
 	},
-	subtractFuelMass(mass) {
-		var tankMass = Object.entries(this.getFuelMass().tankMass);
+	subtractFuelMass: function(mass, shipState) {
+		shipState = shipState || this.designState;
+
+		var tankMass = Object.entries(this.getFuelMass(shipState).tankMass);
 		var tk = 0;
 		var tkId = tankMass[tk][0];
 		var tkMass = tankMass[tk][1];
 
 		while (0 < mass) {
 			// check if tank has necessary fuel
 			if (mass <= tkMass) { // if yes, subtract mass
-				this.designState.objectCache[tkId].state.fullness -= mass/(this.derivedObjects[tkId].baseObject.weightInformation.volumeCapacity * this.derivedObjects[tkId].baseObject.weightInformation.contentDensity);
+				shipState.objectCache[tkId].state.fullness -= mass/(this.derivedObjects[tkId].baseObject.weightInformation.volumeCapacity * this.derivedObjects[tkId].baseObject.weightInformation.contentDensity);
```

Comment:
Why do you access the cache directly? The cache is supposed to be a concern only in the ShipState class. The ShipState class does (or should do) proper checking of whether things are up to date. By using the cache directly, you are assuming that the cached values are up to date, which you can't possibly always know when this method is called.

Reply:
Yes, the cache editing was a consequence of the misunderstanding of the role of `designState`. I updated the cache and the version flags directly to avoid the relevant states from being overridden when the cache is used to perform new calculations. I will remove this kind of handling from `designState` in the future, as you asked. Do you want to meet and talk about this in the following days (weeks)?